testharness/src/test/java/bio/terra/policy/service/pao/PaoUpdateTest.java

```diff
@@ -0,0 +1,487 @@
+    result = paoService.updatePao(paoSid, empty, oneFlag, PaoUpdateMode.FAIL_ON_CONFLICT);
+    assertTrue(result.conflicts().isEmpty());
+    assertTrue(result.computedPao().getEffectiveAttributes().getInputs().isEmpty());
+
+    // Double check that the policy got saved properly
+    checkD = paoService.getPao(paoDid);
+    assertTrue(checkD.getEffectiveAttributes().getInputs().isEmpty());
+  }
+
+  @Test
+  void updatePropagateConflictTest() throws Exception {
+    // Start S1 empty, S2 policy X data2; link to D
+    // add S1 policy X data1; conflict - try with DRY_RUN, FAIL_ON_CONFLICT, ENFORCE_CONFLICT
+    PolicyInput xData2 = makeDataInput(TEST_DATA_POLICY_X, DATA2);
+
+    UUID paoS1id = makePao();
+    logger.info("paoS1id: {}", paoS1id);
+    UUID paoS2id = makePao(xData2);
+    logger.info("paoS2id: {}", paoS2id);
+    UUID paoDid = makePao();
+    logger.info("paoDid: {}", paoDid);
+    paoService.linkSourcePao(paoDid, paoS1id, PaoUpdateMode.FAIL_ON_CONFLICT);
+    paoService.linkSourcePao(paoDid, paoS2id, PaoUpdateMode.FAIL_ON_CONFLICT);
+
+    Pao checkD = paoService.getPao(paoDid);
+    checkForPolicies(checkD, xData2);
+
+    // Conflict case - DRY_RUN
+    PolicyInputs empty = makePolicyInputs();
+    PolicyInputs conflictPolicy = makePolicyInputs(makeDataInput(TEST_DATA_POLICY_X, DATA1));
+
+    PolicyUpdateResult result =
+        paoService.updatePao(paoS1id, conflictPolicy, empty, PaoUpdateMode.DRY_RUN);
+    checkConflict(result, paoDid, paoS2id, new PolicyName(TEST_NAMESPACE, TEST_DATA_POLICY_X));
+    checkD = paoService.getPao(paoDid);
+    checkForPolicies(checkD, xData2);
+    PolicyInput conflicted = checkD.getEffectiveAttributes().getInputs().get(xData2.getKey());
+    assertEquals(0, conflicted.getConflicts().size());
+
+    // Conflict case - FAIL_ON_CONFLICT
+    result = paoService.updatePao(paoS1id, conflictPolicy, empty, PaoUpdateMode.FAIL_ON_CONFLICT);
+    checkConflict(result, paoDid, paoS2id, new PolicyName(TEST_NAMESPACE, TEST_DATA_POLICY_X));
+    checkD = paoService.getPao(paoDid);
+    checkForPolicies(checkD, xData2);
+    conflicted = checkD.getEffectiveAttributes().getInputs().get(xData2.getKey());
+    assertEquals(0, conflicted.getConflicts().size());
+
+    /*
+    TODO[PF-1927] The conflict scheme is not working the way it needs to.
+     Fixing it is outside the scope of this work, but I'll to PF-1927 next!
+    // Conflict case - ENFORCE_CONFLICTS
+    result = paoService.updatePao(paoS1id, conflictPolicy, empty, PaoUpdateMode.ENFORCE_CONFLICTS);
+    checkConflict(result, paoDid, paoS2id, new PolicyName(TEST_NAMESPACE, TEST_DATA_POLICY_X));
+    checkD = paoService.getPao(paoDid);
+    checkForPolicies(checkD, xData2);
+
+    conflicted = checkD.getEffectiveAttributes().getInputs().get(xData2.getKey());
+    assertEquals(1, conflicted.getConflicts().size());
+    assertTrue(conflicted.getConflicts().contains(paoS1id));
+     */
+  }
+
+  private void checkConflict(
```

Comment:
opt In these check methods, sometimes it's not clear what's actual vs expected. This could be:

```
private void checkConflict(
      PolicyUpdateResult actual, UUID expectedDependent, UUID expectedSource, PolicyName expectedPolicyName) {
    assertEquals(1, actual.conflicts().size());
    PolicyConflict actualConflict = actual.conflicts().get(0);
    assertEquals(expectedDependent, actualConflict.dependent().getObjectId());
    assertEquals(expectedSource, actualConflict.source().getObjectId());
    assertEquals(expectedPolicyName, actualConflict.policyName());
  }
```

Reply:
The arguments to the asserts are always (expected, actual). If you don't obey, you get bad error messages.
I'm happy to provide more descriptive names, but I expect test writers to know that and do it properly.